File: maxProfit.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int sz = prices.size();
        if(sz <= 1) return 0;

        int res = 0;
        int m = prices[0];
        for(int i = 1;i < prices.size();i++){
            res = max(res, prices[i] - m);
            m = min(m, prices[i]);
        }
        return res;
    }
};

class Solution2 {
public:
    int maxProfit(vector<int>& prices) {
        int sz = prices.size();
        if(sz <= 1) return 0;

        vector<int> left(sz, 0);
        vector<int> right(sz, 0);
        left[0] = prices[0];
        right[sz-1] = prices[sz-1];
        for(int i = 1; i < sz; i++){
            if(prices[i] < left[i-1]){
                left[i] = prices[i];
            } else {
                left[i] = left[i-1];
            }
        }
        for(int i = sz-2;i >= 0;i--){
            if(prices[i] > right[i+1]){
                right[i] = prices[i];
            } else {
                right[i] = right[i+1];
            }
        }
        int res = 0;
        for(int i = 0; i < sz - 1; i++){
            res = max(res, right[i+1] - left[i]);
        }
        return res;
    }
};

File: tencent/DynamicProgramming/maxProfit.py
# -*- coding: utf-8 -*-
"""

   File Name：     maxProfit
   Author :       jing
   Date：          2020/3/20

   https://leetcode-cn.com/explore/interview/card/tencent/226/dynamic-programming/923/

   买卖股票的最佳时机

   只允许完成一笔交易（即买入和卖出一支股票一次）
"""


class Solution:
    def maxProfit(self, prices) -> int:
        if prices is None or len(prices) < 2:
            return 0
        min_p = prices[0]
        max_profit = 0
        for i in range(1, len(prices)):
            pp = prices[i]
            if pp < min_p:
                min_p = pp
            else:
                max_profit = max(pp - min_p, max_profit)

        return max_profit


if __name__ == '__main__':
    print(Solution().maxProfit([7,1,5,3,6,4]))
